### PR TITLE
Migration of editor component to TypeScript

### PR DESCRIPTION
**Pull request by kreuzersteve**
This pull request contains the first half of migrating the whole application to TypeScript and remove the hybrid approach including all AngularJS leftovers.
The editor module which was migrated contains the following subcomponents:

- editor main component
- editor ace component
- editor file browser component
- editor build output component
- editor instructions component

### CODE REVIEW

File: .idea/compiler.xml [hunk 1]
```diff
@@ -1,13 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (krusche) [inline .idea/compiler.xml]: Please revert the changes in this file!

File: .idea/gradle.xml [hunk 2]
```diff
@@ -1,19 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (krusche) [inline .idea/gradle.xml]: Please revert the changes in this file!

File: .idea/misc.xml [hunk 3]
```diff
@@ -1,12 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (krusche) [inline .idea/misc.xml]: Please revert the changes in this file!

File: .idea/modules.xml [hunk 4]
```diff
@@ -1,11 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (krusche) [inline .idea/modules.xml]: Please revert the changes in this file!

File: .idea/vcs.xml [hunk 5]
```diff
@@ -1,6 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (krusche) [inline .idea/vcs.xml]: Please revert the changes in this file!

File: src/main/webapp/app/courses/results/result.component.ts [hunk 6]
```diff
@@ -1,21 +1,12 @@
...
-import { DomSanitizer } from '@angular/platform-browser';
 import { HttpClient } from '@angular/common/http';
 import { Feedback } from '../../entities/feedback';
 
-@Pipe({name: 'safeHtml'})
-export class SafeHtmlPipe implements PipeTransform {
-    constructor( private sanitizer: DomSanitizer ) { }
-    transform(value) {
-        return this.sanitizer.bypassSecurityTrustHtml(value);
-    }
```
Comment (krusche) [inline src/main/webapp/app/courses/results/result.component.ts]: why did you delete this code?

File: src/main/webapp/app/editor/ace/editor-ace.component.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+    /**
+     * Ace Editor Options
+     */
+    editorText = '';
+    editorFileSessions: object = {};
+    editorMode = 'java'; // String or mode object
+    editorReadOnly = false;
+    editorAutoUpdate = true; // change content when editor text changes
+    editorDurationBeforeCallback = 3000; // wait 3s before callback 'textChanged' sends new value
```
Comment (krusche) [inline src/main/webapp/app/editor/ace/editor-ace.component.ts]: this value might be too high

File: src/main/webapp/app/editor/ace/editor-ace.component.ts [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+            this.editor.getEditor().resize();
+            this.editor._editor.focus();
+        }, err => {
+            console.log('There was an error while getting file', this.fileName, err);
+        });
+    }
+
+    /**
+     * @function saveFile
+     * @desc Saved the currently selected file; is being called when the file is changed (onFileChanged)
```
Comment (krusche) [inline src/main/webapp/app/editor/ace/editor-ace.component.ts]: Saved --> Saves

File: src/main/webapp/app/editor/ace/editor-ace.component.ts [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+            this.onSaveStatusChange({
+                isSaved: false,
+                saveStatusLabel: ' <i class="fa fa-circle-o-notch fa-spin text-info"></i><span class="text-info"> Saving file.</span>'
+            });
+        }
+
+        this.repositoryFileService.update(this.participation.id,
+            fileName,
+            this.editorFileSessions[fileName].code)
+            .debounceTime(3000)
```
Comment (krusche) [inline src/main/webapp/app/editor/ace/editor-ace.component.ts]: please store 3000 as constant in the beginning of the file and give it a descriptive name

File: src/main/webapp/app/editor/ace/editor-ace.component.ts [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    onFileTextChanged(code) {
+        console.log('onFileTextChanged', this.editorFileSessions, this.editorText, this.fileName);
+        if (this.editorFileSessions[this.fileName] !== code && this.editorText !== '') {
+            this.editorFileSessions[this.fileName].code = code;
+            this.editorFileSessions[this.fileName].unsavedChanges = true;
+
+            // Delay saving file by 2 seconds
+            setTimeout(() => {
+                this.saveFile(this.fileName);
+            }, 2500);
```
Comment (krusche) [inline src/main/webapp/app/editor/ace/editor-ace.component.ts]: please store 2500 as constant in the beginning of the file and give it a descriptive name, also make sure the comment is correct (2 vs 2.5 sec)

File: src/main/webapp/app/editor/build-output/editor-build-output.component.html [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+            <span> &nbsp; Build Output</span>
+        </h3>
+    </div>
+    <div class="card-body">
+        <p class="text-center" [hidden]="!isBuilding"><i class="fa fa-circle-o-notch fa-spin"></i> Building...</p>
+        <p class="text-center" [hidden]="!(buildLogs.length == 0) || isBuilding">No build results available</p>
+
+        <dl class="buildoutput dl-horizontal" [hidden]="buildLogs.length == 0 || isBuilding">
+            <ng-container *ngFor="let logEntry of buildLogs">
+                <dt>{{logEntry.time | date:'yyyy-MM-dd HH:mm:ss'}}</dt>
```
Comment (krusche) [inline src/main/webapp/app/editor/build-output/editor-build-output.component.html]: can we use a constant for the date format?

File: src/main/webapp/app/courses/results/result.component.ts [hunk 12]
```diff
@@ -32,13 +23,13 @@ export class SafeHtmlPipe implements PipeTransform {
 export class ResultComponent implements OnInit, OnChanges, OnDestroy {
 
     @Input() participation: Participation;
-    // @Output() onNewResult;
+    @Input() building: boolean;
```
Comment (krusche) [inline src/main/webapp/app/courses/results/result.component.ts]: would be more consistent, if we call it isBuilding (another variable in a different file is also called isBuilding)

File: src/main/webapp/app/courses/results/result.component.ts [hunk 13]
```diff
@@ -32,13 +23,13 @@ export class SafeHtmlPipe implements PipeTransform {
...
+    @Output() newResult = new EventEmitter<object>();
```
Comment (krusche) [inline src/main/webapp/app/courses/results/result.component.ts]: We have result, results and newResult now. This might be too many result variables

File: src/main/webapp/app/courses/results/result.component.ts [hunk 14]
```diff
@@ -60,17 +51,18 @@ export class ResultComponent implements OnInit, OnChanges, OnDestroy {
...
 
             this.init();
-            // make sure result and participation are connected
+            // Make sure result and participation are connected
             if (this.result) {
                 this.result.participation = this.participation;
             }
 
+            // Initial refresh call
+            this.refresh(false);
```
Comment (krusche) [inline src/main/webapp/app/courses/results/result.component.ts]: Are you sure that this code does not introduce breaking changes in other components, e.g. the courses dashboard or the exercise dashboard? I think this code is used in these places as well, so please make 100% sure that it does not introduce problems in the other places!

File: src/main/webapp/app/editor/build-output/editor-build-output.component.ts [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+     * @desc Used to enable resizing for the instructions component
+     */
+    ngAfterViewInit(): void {
+        interact('.resizable-buildoutput')
+            .resizable({
+                // Enable resize from top edge; triggered by class rg-top
+                edges: { left: false, right: false, bottom: false, top: '.rg-top' },
+                // Set min and max height
+                restrictSize: {
+                    min: { height: 50 },
```
Comment (krusche) [inline src/main/webapp/app/editor/build-output/editor-build-output.component.ts]: please use constants for these values

File: src/main/webapp/app/editor/build-output/editor-build-output.component.ts [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+                const target = event.target;
+                // Update element size
+                target.style.width  = event.rect.width + 'px';
+                target.style.height = event.rect.height + 'px';
+        });
+    }
+
+    /**
+     * @function ngOnChanges
+     * @desc Queries for participation results
```
Comment (krusche) [inline src/main/webapp/app/editor/build-output/editor-build-output.component.ts]: provide a more detailed comment, why is this necessary in ngOnChanges?

File: src/main/webapp/app/editor/build-output/editor-build-output.component.ts [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+        }
+    }
+
+    /**
+     * @function getBuildLogs
+     * @desc Gets the buildlogs for the current participation
+     */
+    getBuildLogs() {
+        this.repositoryService.buildlogs(this.participation.id).subscribe( buildLogs => {
+            // TODO: check if buildLogs.log actually exists
```
Comment (krusche) [inline src/main/webapp/app/editor/build-output/editor-build-output.component.ts]: please resolve the TODOs or get rid of them (in case they are resolved)

File: src/main/webapp/app/editor/editor.component.html [hunk 18]
```diff
@@ -1,3 +1,83 @@
...
+    <nav class="navbar navbar-light navbar-expand-md jh-navbar">
+        <div class="jh-logo-container float-left editor-top-left">
+            <a class="logo float-left" routerLink="/">
+                <span jhiTranslate="global.title" class="navbar-title">ArTEMiS:</span> <span>{{participation?.exercise.course.title}}</span>
+            </a>
+        </div>
+        <div class="navbar-collapse collapse editor-top-right">
+            <ul class="navbar-nav ml-auto">
+                <li class="nav-item dropdown pointer">
+                    <!-- TODO: show-score required? -->
```
Comment (krusche) [inline src/main/webapp/app/editor/editor.component.html]: resolve TODO

File: src/main/webapp/app/editor/editor.component.ts [hunk 19]
```diff
@@ -23,44 +25,50 @@ import { NG1WEBSOCKET_SERVICE } from '../shared/websocket/ng1-websocket.service'
...
  */
-export class EditorComponent implements OnInit, OnDestroy {
+export class EditorComponent implements OnInit, AfterViewInit, OnChanges, OnDestroy {
 
-    /** Dependencies as defined by the upgraded Editor component */
+    /** Dependencies as defined by the Editor component */
     participation: Participation;
     repository: RepositoryService;
-    file: any;
+    file: string;
```
Comment (krusche) [inline src/main/webapp/app/editor/editor.component.ts]: good!

File: src/main/webapp/app/editor/editor.component.ts [hunk 20]
```diff
@@ -23,44 +25,50 @@ import { NG1WEBSOCKET_SERVICE } from '../shared/websocket/ng1-websocket.service'
...
      * @desc Framework function which is executed when the component is instantiated.
      * Used to assign parameters which are used by the component
      */
-    ngOnInit() {
+    ngOnInit(): void {
         this.paramSub = this.route.params.subscribe(params => {
             /** Query the participationService for the participationId given by the params */
             this.participationService.find(params['participationId']).subscribe((response: HttpResponse<Participation>) => {
                 this.participation = response.body;
+                console.log('Received a participation!', this.participation);
```
Comment (krusche) [inline src/main/webapp/app/editor/editor.component.ts]: Remove log statements before merging the pull request

File: src/main/webapp/app/editor/editor.component.ts [hunk 21]
```diff
@@ -71,6 +79,143 @@ export class EditorComponent implements OnInit, OnDestroy {
         this.repository = this.repositoryService;
     }
 
+    /**
+     * @function ngAfterViewInit
+     * @desc Used to enable resizing for the instructions component
```
Comment (krusche) [inline src/main/webapp/app/editor/editor.component.ts]: add more detailed comments

File: src/main/webapp/app/editor/editor.component.ts [hunk 22]
```diff
@@ -71,6 +79,143 @@ export class EditorComponent implements OnInit, OnDestroy {
...
+     */
+    ngAfterViewInit(): void {
+        interact('.resizable-filebrowser')
+            .resizable({
+                // Enable resize from right edge; triggered by class rg-right
+                edges: { left: false, right: '.rg-right', bottom: false, top: false },
+                // Set min and max width
+                restrictSize: {
+                    min: { width: 50 },
```
Comment (krusche) [inline src/main/webapp/app/editor/editor.component.ts]: use constants for these values

File: src/main/webapp/app/editor/editor.component.ts [hunk 23]
```diff
@@ -71,6 +79,143 @@ export class EditorComponent implements OnInit, OnDestroy {
...
+        });
+    }
+
+    /**
+     * @function ngOnChanges
+     * @desc Checks if the repository has uncommitted changes
+     * @param changes
+     */
+    ngOnChanges(changes: SimpleChanges) {
+        // TODO: limit this to changes.participation?
```
Comment (krusche) [inline src/main/webapp/app/editor/editor.component.ts]: resolve TODO

File: src/main/webapp/app/editor/editor.component.ts [hunk 24]
```diff
@@ -71,6 +79,143 @@ export class EditorComponent implements OnInit, OnDestroy {
...
+     */
+    updateLatestResult($event) {
+        this.isBuilding = false;
+        this.latestResult = $event.newResult;
+    }
+
+    /**
+     * @function updateSelectedFile
+     * @desc Callback function for when a new file is selected within the file-browser component
+     * @param $event Event object which contains information regarding the ocurred event
```
Comment (krusche) [inline src/main/webapp/app/editor/editor.component.ts]: the description of event is too generic.

File: src/main/webapp/app/editor/editor.component.ts [hunk 25]
```diff
@@ -71,6 +79,143 @@ export class EditorComponent implements OnInit, OnDestroy {
...
+        if (!this.isSaved) {
+            this.isCommitted = false;
+        }
+        this.saveStatusLabel = $event.saveStatusLabel;
+    }
+
+    /**
+     * @function updateLatestResult
+     * @desc Callback function for when a new result is received from the result component
+     * @param $event Event object which contains information regarding the ocurred event
```
Comment (krusche) [inline src/main/webapp/app/editor/editor.component.ts]: the description of event is too generic.

File: src/main/webapp/app/editor/editor.component.ts [hunk 26]
```diff
@@ -71,6 +79,143 @@ export class EditorComponent implements OnInit, OnDestroy {
...
+     */
+    updateSelectedFile($event) {
+        this.file = $event.fileName;
+    }
+
+    /**
+     * @function updateRepositoryCommitStatus
+     * @desc Callback function for when a file was created or deleted; updates the current repository files
+     * @param $event Event object which contains information regarding the ocurred event
```
Comment (krusche) [inline src/main/webapp/app/editor/editor.component.ts]: the description of event is too generic.

File: src/main/webapp/app/editor/editor.service.ts [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * @function getPlantUmlImage
+     * @param plantUml definition obtained by parsing the README markdown file
+     * @desc Requests the plantuml png file as arraybuffer and converts it to base64
+     */
+    @Cacheable({
+        /**
+         * Cacheable configuration
+         */
+        maxCacheCount: 3,
```
Comment (krusche) [inline src/main/webapp/app/editor/editor.service.ts]: please use constants

File: src/main/webapp/app/editor/file-browser/delete-file.html [hunk 28]
```diff
@@ -0,0 +1,18 @@
+<form #deleteForm="ngForm" name="deleteForm" (ngSubmit)="deleteFile()">
+    <div class="modal-header">
+        <h4 class="modal-title">Delete file</h4>
+        <button type="button" class="close" data-dismiss="modal" (click)="closeModal()">&times;</button>
+    </div>
+    <div class="modal-body">
+        <jhi-alert-error></jhi-alert-error>
+            <p class="text-danger">Are you sure you want to delete the file <strong>{{fileNameToDelete}}</strong>?</p>
```
Comment (krusche) [inline src/main/webapp/app/editor/file-browser/delete-file.html]: Are you using the translation service properly?

File: src/main/webapp/app/editor/file-browser/editor-file-browser-create.ts [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+    createFile() {
+        this.isLoading = true;
+        // Make sure we have a filename
+        if (this.newFileName) {
+            const absoluteFilePath = (this.newFileFolder ? this.newFileFolder + '/' : '') + this.newFileName;
+            this.repositoryFileService.create(this.participation.id, absoluteFilePath).subscribe( res => {
+                this.isLoading = false;
+                console.log('Successfully created file: ' + absoluteFilePath, res);
+                this.closeModal();
+                // TODO: select newly created file
```
Comment (krusche) [inline src/main/webapp/app/editor/file-browser/editor-file-browser-create.ts]: resolve TODO

File: src/main/webapp/app/editor/file-browser/editor-file-browser-create.ts [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+     * @desc Reads the provided fileName and folder and creates a new file in the repository
+     */
+    createFile() {
+        this.isLoading = true;
+        // Make sure we have a filename
+        if (this.newFileName) {
+            const absoluteFilePath = (this.newFileFolder ? this.newFileFolder + '/' : '') + this.newFileName;
+            this.repositoryFileService.create(this.participation.id, absoluteFilePath).subscribe( res => {
+                this.isLoading = false;
+                console.log('Successfully created file: ' + absoluteFilePath, res);
```
Comment (krusche) [inline src/main/webapp/app/editor/file-browser/editor-file-browser-create.ts]: remove debugging log statements

File: src/main/webapp/app/editor/file-browser/editor-file-browser-delete.ts [hunk 31]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * @function deleteFile
+     * @desc Reads the provided fileName and deletes the matching file in the repository
+     */
+    deleteFile() {
+        this.isLoading = true;
+        // Make sure we have a filename
+        if (this.fileNameToDelete) {
+            this.repositoryFileService.delete(this.participation.id, this.fileNameToDelete).subscribe( res => {
+                console.log('Successfully deleted file: ' + this.fileNameToDelete, res);
```
Comment (krusche) [inline src/main/webapp/app/editor/file-browser/editor-file-browser-delete.ts]: remove debugging log

File: src/main/webapp/app/editor/file-browser/editor-file-browser.component.html [hunk 32]
```diff
@@ -0,0 +1,32 @@
...
+                  [items]="filesTreeViewItem"
+                  [itemTemplate]="itemTemplate">
+    </ngx-treeview>
+</div>
+
+<ng-template #itemTemplate let-item="item" let-onCollapseExpand="onCollapseExpand" let-onCheckedChange="onCheckedChange">
+    <li *ngIf="item.children"
+        (click)="onCollapseExpand()"
+        class="list-group-item"
+        [style.background-color]="item.checked ? '#428bca' : ''"
```
Comment (krusche) [inline src/main/webapp/app/editor/file-browser/editor-file-browser.component.html]: can we use css here? in particular for colors?

File: src/main/webapp/app/editor/file-browser/editor-file-browser.component.ts [hunk 33]
```diff
@@ -0,0 +1,280 @@
...
+     * @function ngOnInit
+     * @desc Creates the basic configuration for the TreeView (ngx-treeview)
+     */
+    ngOnInit(): void {
+        this.treeviewConfig = TreeviewConfig.create({
+            hasAllCheckBox: false,
+            hasFilter: false,
+            hasCollapseExpand: false,
+            decoupleChildFromParent: false,
+            maxHeight: 1200
```
Comment (krusche) [inline src/main/webapp/app/editor/file-browser/editor-file-browser.component.ts]: why do we have a max height? and in any case, please use a constant

File: src/main/webapp/app/editor/file-browser/editor-file-browser.component.ts [hunk 34]
```diff
@@ -0,0 +1,280 @@
...
+        }
+    }
+
+    /**
+     * @function onCreatedFile
+     * @desc Emmiter function for when a new file was created; notifies the parent component
+     * @param statusChange
+     */
+    onCreatedFile(statusChange: object) {
+        console.log('EMITTING onCreatedFile');
```
Comment (krusche) [inline src/main/webapp/app/editor/file-browser/editor-file-browser.component.ts]: remove debugging log

File: src/main/webapp/app/editor/file-browser/editor-file-browser.component.ts [hunk 35]
```diff
@@ -0,0 +1,280 @@
...
+        this.createdFile.emit(statusChange);
+    }
+
+    /**
+     * @function onDeletedFile
+     * @desc Emmiter function for when a file was deleted; notifies the parent component
+     * @param statusChange
+     */
+    onDeletedFile(statusChange: object) {
+        console.log('EMITTING onDeletedFile');
```
Comment (krusche) [inline src/main/webapp/app/editor/file-browser/editor-file-browser.component.ts]: remove debugging log

File: src/main/webapp/app/editor/instructions/editor-instructions-result-detail.ts [hunk 36]
```diff
@@ -0,0 +1,33 @@
...
+
+// Modal -> Result details view
+@Component({
+    selector: 'jhi-editor-instructions-result-detail',
+    templateUrl: '../../courses/results/result-detail.html'
+})
+export class EditorInstructionsResultDetailComponent implements OnInit {
+    @Input() result: Result;
+    @Input() tests;
+    loading: boolean;
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions-result-detail.ts]: loading --> isLoading

File: src/main/webapp/app/editor/instructions/editor-instructions-result-detail.ts [hunk 37]
```diff
@@ -0,0 +1,33 @@
...
+    filterTests: string;
+    details: Feedback[];
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions-result-detail.ts]: name Missmatch details vs feedback

File: src/main/webapp/app/editor/instructions/editor-instructions-result-detail.ts [hunk 38]
```diff
@@ -0,0 +1,33 @@
...
+    buildLogs;
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions-result-detail.ts]: type is missing for buildLogs

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 39]
```diff
@@ -0,0 +1,434 @@
...
+        ResultService,
+        ParticipationResultService,
+        EditorService
+    ]
+})
+
+export class EditorInstructionsComponent implements OnInit, AfterViewInit, OnChanges, OnDestroy {
+
+    isLoading = false;
+    loadedDetails = false;
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: name 'loadedDetails' is misleading, I would not expect a boolean here

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 40]
```diff
@@ -0,0 +1,434 @@
...
+    initialInstructionsWidth: number;
+    markDown: Remarkable;
+    readMeFileContent: string;
+    readMeFileRendered: string;
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: I guess readMeFileRendered --> readMeFileRenderedContent

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 41]
```diff
@@ -0,0 +1,434 @@
...
+    resultDetails: Feedback[];
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: name mismatch resultDetails vs. feedback

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 42]
```diff
@@ -0,0 +1,434 @@
...
+    steps = [];
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: type for steps is missing

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 43]
```diff
@@ -0,0 +1,434 @@
...
+        });
+    }
+
+    /**
+     * @function setupMarkDown
+     * @desc Initializes the Remarkable object and registers our custom parsing and rendering rules
+     */
+    setupMarkDown() {
+        this.markDown = new Remarkable();
+        // TODO: check if bind required
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: resolve TODO

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 44]
```diff
@@ -0,0 +1,434 @@
...
+     * @param endLine
+     * @param silent
+     */
+    remarkablePlantUmlParser = function(state, startLine, endLine, silent) {
+        const shift = state.tShift[startLine];
+        const max = state.eMarks[startLine];
+        let pos = state.bMarks[startLine];
+        pos += shift;
+
+        if (shift > 3 || pos + 2 >= max) { return false; }
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: I think some additional comments would be good here, what do slight, pos and max mean?

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 45]
```diff
@@ -0,0 +1,434 @@
...
+        if (state.src.charCodeAt(pos) !== 0x40/* @ */) { return false; }
+
+        const ch = state.src.charCodeAt(pos + 1);
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: ch -> char

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 46]
```diff
@@ -0,0 +1,434 @@
...
+            // Call each removal function to detach the click listener from DOM
+            listenerRemoveFunction();
+        }
+        // Set function array to empty
+        this.listenerRemoveFunctions = [];
+    }
+
+    /**
+     * @function remarkablePlantUmlParser
+     * @desc Parser rule for Remarkable custom token PlantUml
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: please add a more detailed description of this function

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 47]
```diff
@@ -0,0 +1,434 @@
...
+            lines: [ startLine, state.line ],
+            content: state.getLines(startLine, state.line, 0, true)
+        });
+
+        return true;
+    };
+
+    /**
+     * @function remarkablePlantUmlRenderer
+     * @desc Renderer rule for Remarkable custom token PlantUml
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: please add a more detailed description of this function

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 48]
```diff
@@ -0,0 +1,434 @@
...
+     * @param tokens
+     * @param id
+     */
+    remarkablePlantUmlRenderer(tokens, id) {
+        let plantUml = tokens[id].content;
+
+        plantUml = plantUml.replace('@startuml', '@startuml\nskinparam shadowing false\nskinparam classBorderColor black\nskinparam classArrowColor black\nskinparam DefaultFontSize 14\nskinparam ClassFontStyle bold\nskinparam classAttributeIconSize 0\nhide empty members\n');
+
+        const that = this;
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: why do we need 'that'?

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 49]
```diff
@@ -0,0 +1,434 @@
...
+        }, err => {
+            console.log('Error getting plantUmlImage', err);
+        });
+
+        return "<img id='plantUml" + id + "' alt='plantUml'" + id + " '/>";
+    }
+
+    /**
+     * @function remarkableTestsStatusParser
+     * @desc Parser rule for Remarkable custom token TestStatus
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: please add a more detailed description of this function

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 50]
```diff
@@ -0,0 +1,434 @@
...
+
+    /**
+     * @function remarkableTestsStatusRenderer
+     * @desc Renderer rule for Remarkable custom token TestStatus
+     * @param tokens
+     * @param id
+     * @param options
+     * @param env
+     */
+    remarkableTestsStatusRenderer(tokens, id: number, options, env) {
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: types are missing for the params

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 51]
```diff
@@ -0,0 +1,434 @@
...
+
+        return text;
+    }
+
+    /**
+     * @function statusForTests
+     * @desc Callback function for renderers to set the appropiate test status
+     * @param tests
+     */
+    statusForTests(tests): object {
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: types are missing for 'tests' and object is not a good return type

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 52]
```diff
@@ -0,0 +1,434 @@
...
+            title: tokens[0].title,
+            done: status['done']
+        });
+
+        return text;
+    }
+
+    /**
+     * @function statusForTests
+     * @desc Callback function for renderers to set the appropiate test status
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: please add a more detailed description of this function

File: src/main/webapp/app/editor/instructions/editor-instructions.component.ts [hunk 53]
```diff
@@ -0,0 +1,434 @@
...
+                    if (result.text === test) {
+                        failedTests++;
+                    }
+                }
+            }
+
+            done = (this.latestResult && this.latestResult.successful) || failedTests === 0;
+            if (totalTests === 1) {
+                if (done) {
+                    label = 'Test passing';
```
Comment (krusche) [inline src/main/webapp/app/editor/instructions/editor-instructions.component.ts]: we should use the translation service here

File: src/main/webapp/app/courses/results/result.component.ts (hunk 6 above)
Comment (kreuzersteve) [reply src/main/webapp/app/courses/results/result.component.ts]: To avoid duplicate declarations: This pipe was used to the shared module (directory shared/pipes), so it can be reused by other components.

File: .idea/compiler.xml (hunk 1 above)
Comment (krusche) [reply .idea/compiler.xml]: test

File: src/main/webapp/app/editor/editor.service.ts (hunk 27 above)
Comment (kreuzersteve) [reply src/main/webapp/app/editor/editor.service.ts]: Need to pass the values to the method decorator directly since it has no reference to the 'this' scope and thereforce can't access local constants.
